### ISSUE TITLE
"Incomplete" status not tied with storage location, leading to inconsistencies?

### ISSUE THREAD

**Issue by cbop-dev** [gilbertchen/duplicacy#188]
This question/issue is related to the question raised in [issue #109](https://github.com/gilbertchen/duplicacy/issues/109): when storing a single repository to more than one storage location, it seems that any backup process, when halted early, is at risk of having its "incomplete" status lost or corrupted by _any other backup process run on the same repository, but to another storage location._

If I run `duplicacy backup -storage aws-location`, for example, but then the program halts early (say, I hit control-C, or there's a power outage), and then run `duplicacy backup` (to the default storage) before starting the `... -storage aws-location` process, will not this running process read the "incomplete" files as though it pertains to itself, ignore supposedly completed files, and then, when finished, either overwrite the "incomplete" file (if it is halted early) or delete it (if it completes)? Won't this result in, at best, an inaccurate incomplete status of a backup snapshot process, or, worse, missing files/objects in the backup storage? Or am I misunderstanding something?

If I am right, doesn't this mean that one should never run concurrent backups of the same source to different backup locations, but must run them serially? (This would make nightly backups, for example, take longer.) Or am I misunderstanding something? Or is the `copy` command supposed to ease this use-case (making serial backups to multiple locations faster)?

But perhaps a simple solution (allowing for concurrent backups to multiple locations) would be to maintain distinct "incomplete" files per storage location: e.g., incomplete/default, incomplete/externalUSBDrive, incomplete/aws, etc., thus allowing each backup process to check and maintain its own complete/incomplete status.

**Comment by jt70471** [comment]
I don't know whether the issue you bring up regarding "incomplete" backups is a problem or not, but thought I'd share the following.

One paradigm of duplicacy that is much different than other backup programs/utilities is that it provides the ability to "copy" a snapshot from one storage destination to another. Instead of backing up your data to multiple destinations, back it up to one location, then use "duplicacy copy" to make a copy of that snapshot.

This has a couple of benefits:

1. The snapshot, once copied, is an exact duplicate of the original backup. If you ran multiple backups, even simultaneously, there would still be differences in the data between the two different backups.

2. There is less overhead in copying a snapshot from one location to another, as opposed to reading through your entire file system, determining whether files have changed or not, etc.

Here's how I'm using duplicacy:

* Backup to a local NAS at GbE speed to -storage nas02
* Copy -from nas02 -to usb01
* Copy -from nas02 -to b2

So, after my backup script runs, I have 2 on-site copies of my data and a 3rd off-site copy stored in the cloud using Backblaze b2.

Hope that helps you when considering how you might handle your backup needs. Good luck!

**Comment by cbop-dev** [comment]
Thanks for the tip about the "copy" command: perhaps that is a better practice for backing up to multiple locations. 

But I wonder: does not the same issue I raised come up with running multiple "copy" commands? That is, is it safe to run, concurrently, two `duplicacy copy` commands, from the same repository, to two different storage locations? (Does duplicacy store the "incomplete" status such commands if it they are halted? If not, does not doing so create a lot of extra overhead when they restart?) I'm thinking, particularly, of creating/uploading an _initial snapshot_ to remote locations, which can take a while for a large repository, and might be nice to get it done as quickly as possible, i.e., run the processes in parallel.

**Comment by jt70471** [comment]
The copy command can be interrupted. Since 2.0.9, duplicacy will check the destination storage to see if a chunk already exists or not and skip copying the chunk. There's no state or incomplete artifacts maintained when copy is processing. If copies or verifies all chunks on the source snapshot exist on the destination, then once that is completed, it then writes the destination snapshot information.

If it's interrupted, it just does the same thing 1) get a list of all chunks in the source 2) list all chunks at the destination 3) skip the ones that already exist and copy the ones that aren't there 4) write destination snapshot id/revision info for the copied snapshots.

**Comment by gilbertchen** [comment]
It may be better for each storage to have a separate incomplete file, but using a single file won't cause incorrectness.  The incomplete file mainly contains a list of files already uploaded as well as their chunks.  When an incomplete file is loaded, the chunks that make up the uploaded files are checked on the storage to make sure they all exists.  If any chunk can't be found the corresponding file will be rescanned again, and only files with chunks that all exist on the storage will be skipped.

**Comment by cbop-dev** [comment]
Thanks, that clarifies things! This system does seem rather robust and less prone to corruption then -- a problem I've found with another product (duplicati) and why I'm looking at duplicacy. (Several times, with duplicati, when an initial backup halted or failed, e.g., a network disconnect, I found the backup status was corrupted and not the database and the server state could not be reconciled, even with a "delete and repair.") 

It still would be nice to have separate "incomplete" status files on a per-storage basis, which would reduce a little overhead when stopping and restarting several parallel backup processes on the same repository. but it is nice to know this is a question of _efficiency,_ not _consistency._

**Comment by gilbertchen** [comment]
Good to know that.  This is a positive side effect of index-less chunk management which leads to simpler and less error-prone implementation.  If you haven't read it, here is a long post by one of our early users in which he explained why Duplicacy's design is advantageous: https://duplicacy.com/issue?id=5651874166341632

https://github.com/gilbertchen/duplicacy/issues/195 has been created to implement per-storage incomplete files.

**Comment by jbrodriguez** [comment]
> Backup to a local NAS at GbE speed to -storage nas02
> Copy -from nas02 -to usb01
> Copy -from nas02 -to b2

@jt70471 do you run the script from your workstation ?

What I mean is duplicacy copy is run on your workstation or on the NAS ?

**Comment by jt70471** [comment]
Hi, the duplicacy copy command is run on my iMac. The usb01 external USB drive is only attached to my iMac. I could attach that drive to my NAS and run duplicacy from there of course, but I do a lot of scripting from my iMac, which is on all of time...
